src/backend/dx12/src/command.rs

```diff
@@ -198,19 +311,119 @@ impl CommandBuffer {
             );
         }
     }
+
+    fn set_gr_bind_point(&mut self) {
+        if self.active_bindpoint != BindPoint::Graphics {
+            // Switch to graphics bind point
+            assert!(self.gr_pipeline.pipeline.is_some(), "No graphics pipeline bound");
+            let (pipeline, _) = self.gr_pipeline.pipeline.unwrap();
+            unsafe { self.raw.SetPipelineState(pipeline); }
+            self.active_bindpoint = BindPoint::Graphics;
+        }
+
+        self.flush_gr_user_data();
+    }
+
+    fn flush_gr_user_data(&mut self) {
+        let user_data = &mut self.gr_pipeline.user_data;
+        if user_data.dirty_mask == 0 {
+            return
+        }
+
+        // TODO: root constant support
+
+        // Flush descriptor tables
+        // Index in the user data array where tables are starting
+        let table_start = 0;
+        let table_slot_start = 0;
+        for i in 0..self.gr_pipeline.num_parameter_slots {
+            let table_index = i + table_start;
+            if ((user_data.dirty_mask >> table_index) & 1) == 1 {
+                let slot = (i + table_slot_start) as _;
+                match user_data.data[i] {
+                    RootElement::TableSrvCbvUav(offset) => {
+                        let gpu = winapi::D3D12_GPU_DESCRIPTOR_HANDLE {
+                            ptr: self.gr_pipeline.srv_cbv_uav_start + offset as u64,
+                        };
+                        unsafe { self.raw.SetGraphicsRootDescriptorTable(slot, gpu); }
+                    },
+                    RootElement::TableSampler(offset) => {
+                        let gpu = winapi::D3D12_GPU_DESCRIPTOR_HANDLE {
+                            ptr: self.gr_pipeline.sampler_start + offset as u64,
+                        };
+                        unsafe { self.raw.SetGraphicsRootDescriptorTable(slot, gpu); }
+                    },
+                    _ => {
+                        error!("Unexpected user data element in the root signature ({:?})", user_data.data[i]);
+                    },
+                }
+                user_data.clear(table_index);
+            }
+        }
+    }
+
+    fn set_compute_bind_point(&mut self) {
+        if self.active_bindpoint != BindPoint::Compute {
+            // Switch to compute bind point
+            assert!(self.comp_pipeline.pipeline.is_some(), "No compute pipeline bound");
+            let (pipeline, _) = self.comp_pipeline.pipeline.unwrap();
+            unsafe { self.raw.SetPipelineState(pipeline); }
+            self.active_bindpoint = BindPoint::Compute;
+        }
+
+        self.flush_compute_user_data();
+    }
+
+    fn flush_compute_user_data(&mut self) {
```

Comment:
seems to be awfully similar to `flush_gr_user_data`. Would it make sense to share the implementations?

Reply:
hmm, I can try to unify them (I remember that we have something similar done prior).